packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,453 @@
+      Promise<Record<SnapId, Record<NamespaceId, Namespace>>>
+    >(async (previousPromise, snap) => {
+      const acc = await previousPromise;
+      const snapNamespaces = await this.snapToNamespaces(snap);
+      if (snapNamespaces !== null) {
+        return { ...acc, [snap.id]: snapNamespaces };
+      }
+      return acc;
+    }, Promise.resolve({}));
+
+    // The magical matching algorithm specified in SIP-2
+    const namespaceToSnaps = findMatchingKeyringSnaps(
+      connection.requiredNamespaces,
+      availableNamespaces,
+    );
+
+    const permissions = await this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      origin,
+    );
+
+    const approvedNamespacesAndSnaps = Object.entries(namespaceToSnaps).reduce(
+      (acc, [namespace, snapIds]) => {
+        // If snap already is approved for use, solve conflict by using that snap
+        const approvedSnap = snapIds.find((snapId) => {
+          return (
+            permissions &&
+            hasProperty(permissions, getSnapPermissionName(snapId))
+          );
+        });
+
+        if (approvedSnap) {
+          return { ...acc, [namespace]: [approvedSnap] };
+        }
+        return acc;
+      },
+      {},
+    );
+
+    const hasConflicts = Object.keys(namespaceToSnaps).some(
+      (namespace) => !hasProperty(approvedNamespacesAndSnaps, namespace),
+    );
+
+    const filteredNamespacesAndSnaps = hasConflicts
+      ? namespaceToSnaps
+      : approvedNamespacesAndSnaps;
+
+    const possibleAccounts = await this.namespacesToAccounts(
+      origin,
+      filteredNamespacesAndSnaps,
+      connection.requiredNamespaces,
+    );
+
+    // If current configuration doesn't solve request, we need to show a prompt.
+    const resolvedAccounts = hasConflicts
+      ? await this.resolveConflicts(origin, possibleAccounts)
+      : possibleAccounts;
+    const chosenAccounts = fromEntries(
+      Object.entries(resolvedAccounts).map(([namespace, snapIds]) => [
+        namespace,
+        snapIds[0] ?? null,
+      ]),
+    );
+
+    const approvedPermissions = Object.values(chosenAccounts).reduce(
+      (acc, cur) => {
+        if (cur) {
+          return {
+            ...acc,
+            [getSnapPermissionName(cur.snapId)]: {},
+          };
+        }
+        return acc;
+      },
+      {},
+    );
+    // TODO: Save the approved permissions
+    await this.messagingSystem.call('PermissionController:grantPermissions', {
+      approvedPermissions,
+      subject: { origin },
+    });
+
+    const providedNamespaces: Record<NamespaceId, SessionNamespace> =
+      fromEntries(
+        Object.entries(connection.requiredNamespaces)
+          .filter(([namespaceId]) => chosenAccounts[namespaceId] !== null)
+          .map(([namespaceId, namespace]) => [
+            namespaceId,
+            {
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+              accounts: chosenAccounts[namespaceId]!.accounts,
```

Comment:
This should help TypeScript without needing to ignore linting
```suggestion
              assert(chosenAccounts[namespaceId] !== null);
              accounts: chosenAccounts[namespaceId]!.accounts,
```

Reply:
Seems unnecessary to run this given that we know it wont fail?